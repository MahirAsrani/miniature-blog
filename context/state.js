import axios from 'axios';
import { useRouter } from 'next/router';
import { createContext, useEffect, useState } from 'react';
import { GET_AUTH, GET_CAT, GET_LOGOUT } from '../utils/constants';

export const AppContext = createContext();

export function ContextWrapper({ children }) {
  const [user, setUser] = useState(null);
  const [loading, setLoading] = useState(false);
  const [category, setCategory] = useState([]);
  const Router = useRouter();

  async function fetchCategory() {
    try {
      const res = await axios.get(GET_CAT);
      res.data.success && setCategory(res.data.data);
    } catch (error) {
      setCategory([]);
    }
  }
  async function fetchUser() {
    try {
      const res = await axios.get(GET_AUTH, {
        withCredentials: true,
      });
      res.data.success && setUser(res.data.data);
    } catch (error) {
      setUser(null);
    }
  }

  useEffect(() => {
    fetchUser();
    fetchCategory();
  }, []);

  async function signOut() {
    await axios.get(GET_LOGOUT, {
      withCredentials: true,
    });

    setUser(null);
    Router.push('/');
  }

  let state = { user, signOut, fetchUser, loading, category };

  return <AppContext.Provider value={state}>{children}</AppContext.Provider>;
}
