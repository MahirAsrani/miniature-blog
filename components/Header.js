import Link from 'next/link';
import { useRouter } from 'next/router';
import React, { useContext } from 'react';
import { AppContext } from '../context/state';

const Header = () => {
  const { user, signOut, category } = useContext(AppContext);
  const Router = useRouter();

  return (
    <div className="container mt-3">
      <nav className="navbar navbar-expand-lg navbar-light">
        <div className="container-fluid">
          <img
            src="https://i.imgur.com/hSDDP67.png"
            height="50px"
            width="50px"
            alt=""
            onClick={() => Router.push('/')}
          />

          <button
            className="navbar-toggler"
            type="button"
            data-bs-toggle="collapse"
            data-bs-target="#navbarSupportedContent"
            aria-controls="navbarSupportedContent"
            aria-expanded="false"
            aria-label="Toggle navigation"
          >
            <span className="navbar-toggler-icon"></span>
          </button>

          <div className="collapse navbar-collapse" id="navbarSupportedContent">
            <ul className="navbar-nav ms-auto mx-4 mb-lg-0">
              <li className="nav-item">
                <Link href="/">
                  <a
                    className={`nav-link ${
                      Router?.pathname === '/' && 'active'
                    }`}
                  >
                    <span className="home">Home</span>
                  </a>
                </Link>
              </li>

              <li className="nav-item dropdown">
                <a
                  className="nav-link dropdown-toggle"
                  id="navbarDropdown"
                  role="button"
                  data-bs-toggle="dropdown"
                  aria-expanded="false"
                >
                  Category
                </a>
                <ul className="dropdown-menu" aria-labelledby="navbarDropdown">
                  {category.map((c) => (
                    <li key={c._id}>
                      <Link href={`/${c.slug}`}>
                        <a className="dropdown-item">{c.title}</a>
                      </Link>
                    </li>
                  ))}{' '}
                  {/* <li>
                    <hr className="dropdown-divider" />
                  </li> */}
                </ul>
              </li>

              {user && (
                <li className="nav-item ">
                  <Link href="/dashboard">
                    <a
                      className={`nav-link ${
                        Router?.pathname === '/dashboard' && 'active'
                      }`}
                    >
                      Dashboard
                    </a>
                  </Link>
                </li>
              )}
            </ul>
            {user ? (
              <div className="btn btn-dark" onClick={signOut}>
                Sign Out
              </div>
            ) : (
              <div
                className="btn btn-dark"
                onClick={() => Router.push('/login')}
              >
                Sign In
              </div>
            )}
          </div>
        </div>
      </nav>
    </div>
  );
};

export default Header;
