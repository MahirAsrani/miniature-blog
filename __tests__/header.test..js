import { render, screen } from '@testing-library/react';
import Header from '../components/Header';
import { ContextWrapper } from '../context/state';

describe('Header', () => {
  it('renders header unchanged', () => {
    const { container } = render(
      <ContextWrapper>
        <Header />
      </ContextWrapper>
    );
    expect(container).toMatchSnapshot();
  });
});
