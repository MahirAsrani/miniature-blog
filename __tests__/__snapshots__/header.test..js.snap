// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header renders header unchanged 1`] = `
<div>
  <div
    class="container mt-3"
  >
    <nav
      class="navbar navbar-expand-lg navbar-light"
    >
      <div
        class="container-fluid"
      >
        <img
          alt=""
          height="50px"
          src="https://i.imgur.com/hSDDP67.png"
          width="50px"
        />
        <button
          aria-controls="navbarSupportedContent"
          aria-expanded="false"
          aria-label="Toggle navigation"
          class="navbar-toggler"
          data-bs-target="#navbarSupportedContent"
          data-bs-toggle="collapse"
          type="button"
        >
          <span
            class="navbar-toggler-icon"
          />
        </button>
        <div
          class="collapse navbar-collapse"
          id="navbarSupportedContent"
        >
          <ul
            class="navbar-nav ms-auto mx-4 mb-lg-0"
          >
            <li
              class="nav-item"
            >
              <a
                class="nav-link false"
                href="/"
              >
                <span
                  class="home"
                >
                  Home
                </span>
              </a>
            </li>
            <li
              class="nav-item dropdown"
            >
              <a
                aria-expanded="false"
                class="nav-link dropdown-toggle"
                data-bs-toggle="dropdown"
                id="navbarDropdown"
                role="button"
              >
                Category
              </a>
              <ul
                aria-labelledby="navbarDropdown"
                class="dropdown-menu"
              >
                 
              </ul>
            </li>
          </ul>
          <div
            class="btn btn-dark"
          >
            Sign In
          </div>
        </div>
      </div>
    </nav>
  </div>
</div>
`;
